crates/table/src/layout.rs

```diff
@@ -440,19 +440,96 @@ impl AlgebraicTypeLayout {
     pub fn algebraic_type(&self) -> AlgebraicType {
         match self {
             AlgebraicTypeLayout::Primitive(prim) => prim.algebraic_type(),
-            AlgebraicTypeLayout::VarLen(var_len) => var_len.algebraic_type(),
+            AlgebraicTypeLayout::VarLen(var_len) => var_len.algebraic_type().clone(),
             AlgebraicTypeLayout::Product(prod) => AlgebraicType::Product(prod.product_type()),
             AlgebraicTypeLayout::Sum(sum) => AlgebraicType::Sum(sum.sum_type()),
         }
     }
+
+    /// Check that this `AlgebraicTypeLayout` is a sub-layout of `other`.
+    ///
+    /// A layout L is a sub-layout of L' if any value deserialized to `AlgebraicValue` via L
+    /// can also be deserialized to an identical `AlgebraicValue` via L'. We additionally require
+    /// that the names in `L` and `L'` match.
+    ///
+    /// This means that:
+    ///
+    /// - `self.size() == other.size()`
+    ///
+    /// - `self.align() >= other.align()` (in other words, `other.align()` divides `self.align()`)
+    ///
+    /// - If `self` is a product type, then:
+    ///     - `other` must also be a product type.
```

Comment:
This is a stronger requirement than we actually need - one-element products are representationally equivalent to their contents. But I don't see any downside to imposing this restriction artificially.